File: Algorithms/110. Balanced Binary Tree/python/110_balanced_binary_tree.py
from tabnanny import check
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        
        left_height = self.checkHeight(root.left)
        right_height = self.checkHeight(root.right)
        
        if abs(left_height - right_height) <=1:
            return self.isBalanced(root.left) and self.isBalanced(root.right)
        else:
            return False
        
    
    def checkHeight(self, root):
        if not root:
            return 0
        left_height = self.checkHeight(root.left)
        right_height = self.checkHeight(root.right)
        
        return 1 + max(left_height, right_height)
        
    
sol = Solution()

test1 = TreeNode(3,
                 TreeNode(9),
                 TreeNode(20,
                          TreeNode(15),
                          TreeNode(7)))

print(sol.isBalanced(test1))

File: Algorithms/110. Balanced Binary Tree/go/110_balanced_binary_tree.go
package main

import (
	"fmt"
)

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isBalanced(root *TreeNode) bool {

	if root == nil {
		return true
	}

	lh := height(root.Left)
	rh := height(root.Right)

	res := lh - rh

	if res < 0 {
		res = res * -1
	}

	if res <= 1 {
		return isBalanced(root.Left) && isBalanced(root.Right)
	}

	return false
}

func height(root *TreeNode) int {
	if root == nil {
		return 0
	}

	lh := height(root.Left)
	rh := height(root.Right)

	if lh < rh {
		return rh + 1
	} else {
		return lh + 1
	}
}

func main() {

	test1 := &TreeNode{Val: 3,
		Left: &TreeNode{Val: 9, Left: nil, Right: nil},
		Right: &TreeNode{Val: 20,
			Left:  &TreeNode{Val: 15, Left: nil, Right: nil},
			Right: &TreeNode{7, nil, nil}},
	}

	fmt.Printf("Result: %v\n", isBalanced(test1))
}
